python/perf-kernels/flash-attention.py

```diff
@@ -543,7 +663,12 @@ def forward(ctx, q, k, v, causal, bias, sm_scale, split_kernel=False):
         num_warps = 8 if Lq == 128 else 4
         pre_load_v = False if Lq == 128 else True
         grid = (triton.cdiv(q.shape[2], BLOCK_M), q.shape[0] * q.shape[1], 1)
-        
+
+        if return_encoded_softmax:
+            encoded_softmax = torch.zeros((q.shape[0], q.shape[1], q.shape[2], k.shape[2]), device=q.device, dtype=torch.float32)
```

Comment:
torch.empty is faster.

Reply:
Anything related to `encoded_softmax` is for debugging and validation, and filling zeros is intentional.
We need to make sure there were actual writes to `encoded_softmax`, and torch.empty nullified this purpose.

Reply:
I cannot comprehend the response, because I do not understand what encoded_softmax is for. It would be helpful to address my comment above on adding information about its purpose.

For example, I'm not sure how filling zeros helps. p can be zeros as well. Or are we relying on the fact that p won't be zeros because for it to be 0, qk must be inf and that would indicate a problem? It is not clear.

Reply:
softmax is literally `softmax(Q'K)`. `encoded_softmax` set the sign bit according to the `dropout_mask`.
We need this information to validate the correctness of dropout operations.

filling zero is a way to validate the chuck of memory is updated. If using `torch.empty`, how can you tell the data was actually written by the kernel, or simply some random garbage filled by a previous user of this chunk of memory?

Again, this variable is for debugging. Arguing with performance is nonsense.

Reply:
It would be helpful if you can please take a step back and try to understand what I am trying to get at.

My original comment is borne out of a lack of understanding of this feature. I am not trying to argue about anything, except may be for adding a comment somewhere here on what this thing is. I do not think that is too much to ask for, and I do not appreciate that being called nonsense.

You did add that information here, and it does help me understand it now, but it could have been quicker and more peaceful to just add the damn comment in the code.

Reply:
Let me add more context to my requests. The Triton team is responsible for this kernel. If there are issues here further down, that is on us. For us to be able to fix those issues, we need to understand all features included herein, including their purpose. 

I think we've been pretty open to anything that helps or makes life easy for AOT w.r.t this kernel and codebase. So I don't think it is too much to ask for details on new features being added here that we've not seen before, nor is available in the literature.